### PR TITLE
UnexpecTEDIntentPolicy: Deduplicate multiple data points with same input state

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Remove the need for creating multiple data points with duplicate input state but permuted version of positive label ids.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by dakshvar22** [comment]
@kedz @JEM-Mosig PR ready for review :)

File: rasa/core/featurizers/tracker_featurizers.py
Comment (JEM-Mosig) [inline rasa/core/featurizers/tracker_featurizers.py:1018]: In the comment it says "list of ints", but the type annotation says "List[List[Text]]".
```suggestion
        # the state. Note that each individual 'label' instance is a list of strings.
```

File: rasa/core/featurizers/tracker_featurizers.py
Comment (JEM-Mosig) [inline rasa/core/featurizers/tracker_featurizers.py:1076]: Can you please add a comment explaining why `[0]`?

**Review by JEM-Mosig — APPROVED** [review]

File: rasa/core/featurizers/tracker_featurizers.py
Comment (dakshvar22) [reply rasa/core/featurizers/tracker_featurizers.py:1079]: Done

**Comment by dakshvar22** [comment]
@kedz I've addressed Johannes' comments but will wait for an approval from you too before I merge.

File: rasa/core/featurizers/tracker_featurizers.py
Comment (kedz) [inline rasa/core/featurizers/tracker_featurizers.py:1052]:
```suggestion

                if self.remove_duplicates:
                    hashed = self._hash_example(states, label)
                    if hashed in hashed_examples:
                        continue
                    hashed_examples.add(hashed)

                # Store all positive labels associated with a training state.
                state_hash = self._hash_example(states)
                state_hash_to_label_list_instances[state_hash].add(label[0])

                # Only add unique example states unless `remove_duplicates` is `False`.
                if not self.remove_duplicates or state_hash not in seen_states:
                    example_states.append(states)
                    example_entities.append(entities)
                    seen_states.add(state_hash)

                pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_states):d}"})

        # Collect positive labels for each state example.
        example_labels = [
            list(state_hash_to_label_list_instances[self._hash_example(state)])
            for state in example_states
        ]

        self._remove_user_text_if_intent(example_states)

        logger.debug(
            f"Created {len(example_states)} {self.LABEL_NAME} examples."
        )

        return example_states, example_labels, example_entities
```

(I can't extend the suggestion after line 1056, but this would replace everything from 1041 on in the function.)

What you have from lines 1058 on works, but IMO it seems wasteful of resources to (a) create both a large list of possibly redundant data and then a pruned list when we could just create the pruned list and (b) create `example_labels` when it isn't being used for anything.

The above suggestion would create pruned example states in one pass and simplify the creation of the pruned example labels later. Not necessary to do but I thought if we have a little more time since we have a release blocking bug we could simplify this code a bit.

This would also mean moving the creation of `seen_states` up to line 1022 and converting state_hash_to_label_list_instances to hold a set instead of a list of length 1 lists (and a name change).

If you don't do this suggestion, I still think you should remove `example_labels` since its not doing anything.

File: rasa/utils/tensorflow/layers_utils.py
Comment (kedz) [inline rasa/utils/tensorflow/layers_utils.py:93]: Why is this only reduced along the last axis when a mask is passed?

File: tests/test_memory_leak.py
Comment (kedz) [inline tests/test_memory_leak.py:159]: Shouldn't we be using less memory now? Or is this unrelated to deduplication of states?

File: rasa/utils/tensorflow/layers_utils.py
Comment (dakshvar22) [reply rasa/utils/tensorflow/layers_utils.py:93]: Hmm, good question. There is an assumption that the actual mask values are on the last dimension and hence the first `reduce_mean` is needed only on that axis. The next `reduce_mean` is across all axis.

File: tests/test_memory_leak.py
Comment (dakshvar22) [reply tests/test_memory_leak.py:159]: This is unrelated. This creeped into the PR from the main PR.

File: rasa/utils/tensorflow/layers_utils.py
Comment (kedz) [inline rasa/utils/tensorflow/layers_utils.py:93]: As advertised in the docstring and name, this function seems agnostic to the dimensions, yet when a mask is passed in it suddenly treats the last dimension as the one of concern and makes everything else a batch dimension over which it then averages. This gives counter intuitive behavior. Based on the behavior of the unmasked case, I would expect
```python
x = [[1,2,3,4]
     [5,6,7,8]] 
y = [[1,2,3,4]
     [5,6,0,0]]
mask = [[1,1,1,1],
        [1,1,1,0]]
```
to give `6 / 7` since there are 6 total matches on 7 unmasked items.
What is currently computed however is  `((4/4) + 2/3) / 2`
             
Perhaps we should add an axis argument that gets passed so you can specify the desired behavior explicitly similar to the underlying tensorflow ops `reduce_mean` and `reduce_sum` upon which this operation is based?

File: rasa/core/featurizers/tracker_featurizers.py
Comment (kedz) [inline rasa/core/featurizers/tracker_featurizers.py:1020]:
```suggestion
        # Store of example hashes (of both states and labels) for removing 
        # duplicate training examples.
        hashed_examples = set()
        # Mapping of example state hash to set of positive labels associated with
        # the state. Note that each individual 'label' instance is a set of strings.
        state_hash_to_label_set: defaultdict[int, Set[Text]] = defaultdict(set)
        # Store of example state hashes so only unique states are kept when 
        # removing duplicates.
        seen_states = set()
```

Added a comment for `seen_states` and modified comments of `state_hash_to_label_set`.

File: rasa/core/featurizers/tracker_featurizers.py
Comment (kedz) [inline rasa/core/featurizers/tracker_featurizers.py:1055]:
```suggestion
                state_hash = self._hash_example(states)

                # Only add unique example states unless `remove_duplicates` is `False`.
                if not self.remove_duplicates or state_hash not in state_hash_to_label_set:
                    example_states.append(states)
                    example_entities.append(entities)

                # Store all positive labels associated with a training state.
                state_hash_to_label_set[state_hash].add(label[0])
```
Sorry last comment on this section I promise :) . I just realized we could get rid of `seen_states` completely since we are already keeping the same information in `state_hash_to_label_set`.

**Review by kedz — CHANGES_REQUESTED** [review]
Tracker featurization code is looking good -- added a suggestion for comment fix and an optional tweak.

Have some questions/concerns about the some of the tensorflow helper functions though.

File: rasa/core/featurizers/tracker_featurizers.py
Comment (dakshvar22) [reply rasa/core/featurizers/tracker_featurizers.py:1020]: Omitting ```Note that each individual 'label' instance is a set of strings.``` since that just mentions the data type which is already known by the type annotation.

File: rasa/utils/tensorflow/layers_utils.py
Comment (dakshvar22) [reply rasa/utils/tensorflow/layers_utils.py]: I clarified it in the docstrings for now. We should give it a thought on how to make it more general purpose in the future (adding the axis argument).

**Comment by dakshvar22** [comment]
/modeltest

```yml

include:
 - dataset: ["all"]
   config: ["all"]

```

File: rasa/utils/tensorflow/layers_utils.py
Comment (kedz) [reply rasa/utils/tensorflow/layers_utils.py]: I'll create an issue for it.

**Review by kedz — APPROVED** [review]
Looks good!

**Pull request merged**
